devito/finite_differences/differentiable.py

```diff
@@ -0,0 +1,236 @@
+import sympy
+import numpy as np
+
+from devito.symbolics.search import retrieve_functions
+from devito.symbolics.extended_sympy import FrozenExpr
+
+__all__ = ['Differentiable']
+
+
+class Differentiable(FrozenExpr):
+    """
+    This class represents Devito differentiable objects such as
+    sum of functions, product of function or FD approximation and
+    provides FD shortcuts for such expressions
+    """
+    _op_priority = 100.0
+    is_Function = True
+    is_Indexed = False
+
+    def __new__(cls, *args, **kwargs):
+        return sympy.Expr.__new__(cls, *args)
+
+    def __init__(self, expr, **kwargs):
+        from devito.finite_differences.finite_difference import generate_fd_functions
+        self.expr = expr.expr if isinstance(expr, Differentiable) else expr
+        self.indices = self._indices()
+        # Generate FD shortcuts for expression or copy from input
+        if isinstance(expr, Differentiable):
+            self.fd = expr.fd
+            # Recover the list of possible FD shortcuts
+            self.dtype = expr.dtype
+            self.space_order = expr.space_order
+            self.time_order = expr.time_order
+            self.staggered = expr.staggered
+        else:
+            # Recover the list of possible FD shortcuts
+            self.dtype = self._dtype()
+            self.space_order = self._space_order()
+            self.time_order = self._space_order()
+            self.staggered = self._staggered()
+            self.fd = kwargs.get('fd', generate_fd_functions(self))
+
+        for d in self.fd:
+            setattr(self.__class__, d[1], property(d[0], d[1]))
+        self.derivatives = tuple(d[1] for d in self.fd)
+
+    def __add__(self, other):
+        if isinstance(other, Differentiable):
+            return Differentiable(sympy.Add(*[self.expr, other.expr], evaluate=False),
+                                  fd=self.fd)
+        else:
+            return Differentiable(sympy.Add(*[self.expr, other], evaluate=False),
+                                  fd=self.fd)
+
+    __radd__ = __add__
+
+    def __iadd__(self, other):
+        if isinstance(other, Differentiable):
```

Comment:
instead of this painful to read cascade of
``if isinstance(other, Differentiable) ...``
I suggest to use a decorator like here:
https://github.com/opesci/devito/blob/master/devito/ir/support/basic.py#L63
in the decorator you can "normalize" ``other`` so that the method (__iadd__, __add__, and all others...) always gets ``expr``.
This little change should halve the size of this file :grimacing:

Reply:
Will look at it